src/Microsoft.Extensions.Hosting.Abstractions/BackgroundService.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Hosting
+{
+    /// <summary>
+    /// Base class for implementing a long running <see cref="IHostedService"/>.
+    /// </summary>
+    public abstract class BackgroundService : IHostedService
+    {
+        private Task _executingTask;
+        private CancellationTokenSource _cts;
+
+        /// <summary>
+        /// This method is called when the <see cref="IHostedService"/> starts. The implementation should return a task that represents
+        /// the lifetime of the long running operation(s) being performed.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> that fires when the <see cref="IHostedService"/> is stopped or cancelled.</param>
+        /// <returns>A <see cref="Task"/> that represents the long running operations.</returns>
+        protected abstract Task ExecuteAsync(CancellationToken cancellationToken);
+
+        public Task StartAsync(CancellationToken cancellationToken)
+        {
+            // If the token is cancellable, create a linked token so we can trigger cancellation 
+            // on shutdown and based on the incoming token
+            _cts = cancellationToken.CanBeCanceled ?
+                CancellationTokenSource.CreateLinkedTokenSource(cancellationToken) :
+                new CancellationTokenSource();
+
+            // Store the task we're executing
+            _executingTask = ExecuteAsync(_cts.Token);
```

Comment:
This doesn't honor the spirit of StartAsync, it should block until the service has successfully started. This pattern readily doesn't differentiate between starting and running. The cancellation tokens share share the same confusion.

Reply:
There's no easy way to indicate that other than introducing some other control flow mechanic (Like a callback). Is it really worth it? It's similar to the 2 token thing.